### PR TITLE
Introduce `--codegen` argument so targets can be specified explicitly

### LINKED ISSUE #3196: Add flag to disable Javascript backend
I'm writing an alternative VimL backend for PureScript based on the CoreFn dump produced by `--dump-corefn`. That is, I haven't forked the compiler codebase. There are two things with the regular `purs compile` command that would be nice to disable:

* CommonJS module output
* Checking/inclusion of `<Module>.js` files (foreign definitions)

Specifically the latter is of interest, as I don't want to write foreign definitions for Vim-specific things for the JS backend.

At a high level, this might be considered "disabling the entire JS backend", I'd say, so I'm not sure what would be a good name for a flag to disable it. Perhaps `--no-codegen` or `--no-js-codegen`. Anyway, I could have a stab at this and provide a PR if this seems reasonable.

Thanks!

### PR DESCRIPTION
**Pull request by garyb**
Resolves #3196, closes #3258 

Default is `js`, using `sourcemaps` implies `js` and the separate option for source maps is gone now.

Thanks to @gabejohnson for doing the first part of this!

### CODE REVIEW

**Comment by garyb** [comment]
Any thoughts/comments about this @kRITZCREEK @LiamGoodacre @hdgarrood?

File: app/Command/Compile.hs [hunk 1]
```diff
@@ -122,21 +126,45 @@ jsonErrors = Opts.switch $
...
+  , ("corefn", P.CoreFn)
+  ]
+
+targetsMessage :: String
+targetsMessage = "Accepted codegen targets are '" <> intercalate "', '" (M.keys targets) <> "'."
+
+targetParser :: Opts.ReadM [P.CodegenTarget]
+targetParser =
+  Opts.str >>= \s ->
+    for (T.split (== ',') s)
```
Comment (kRITZCREEK) [inline app/Command/Compile.hs]: Do we want to be a bit more lenient here and strip the text after splitting?

**Comment by hdgarrood** [comment]
:+1: I like this a lot!

File: app/Command/Compile.hs (hunk 1 above)
Comment (garyb) [reply app/Command/Compile.hs]: I'm not sure how you'd get spaces into that value anyway? Since `--codegen js, corefn` wouldn't parse as one thing.

Comment (garyb) [reply app/Command/Compile.hs]: Oh... quotes maybe?

**Comment by LiamGoodacre** [comment]
Awesome.  No comments other than what @kRITZCREEK raised. :100:

File: app/Command/Compile.hs [hunk 2]
```diff
@@ -154,6 +154,8 @@ targetParser =
       $ maybe (Opts.readerError targetsMessage) pure
       . flip M.lookup targets
       . T.unpack
+      . T.takeWhile (/= ' ')
```
Comment (kRITZCREEK) [inline app/Command/Compile.hs]: How about using http://hackage.haskell.org/package/text-1.2.3.0/docs/Data-Text.html#v:strip?

Comment (garyb) [reply app/Command/Compile.hs]: Thanks, couldn't find that when I looked through the docs before!

**Pull request merged**